--- old/src/share/vm/runtime/deoptimization.hpp	2016-08-15 16:43:35.676545495 +0200
+++ new/src/share/vm/runtime/deoptimization.hpp	2016-08-15 16:43:35.572545495 +0200
@@ -135,7 +135,14 @@
   static void deoptimize(JavaThread* thread, frame fr, RegisterMap *reg_map);
   static void deoptimize(JavaThread* thread, frame fr, RegisterMap *reg_map, DeoptReason reason);
 
+  static void print_deoptimization_count();
+
   private:
+
+  static volatile int _deoptimization_count;
+
+  static void increase_deoptimization_count();
+
   // Does the actual work for deoptimizing a single frame
   static void deoptimize_single_frame(JavaThread* thread, frame fr, DeoptReason reason);
 
