--- old/src/share/vm/runtime/java.cpp	2016-08-15 16:52:16.440545495 +0200
+++ new/src/share/vm/runtime/java.cpp	2016-08-15 16:52:16.356545495 +0200
@@ -312,6 +312,10 @@
     CodeCache::print();
   }
 
+  if (PrintDeoptimizationCount) {
+    Deoptimization::print_deoptimization_count();
+  }
+
   if (PrintMethodFlushingStatistics) {
     NMethodSweeper::print();
   }
@@ -371,6 +375,10 @@
     CodeCache::print();
   }
 
+  if (PrintDeoptimizationCount) {
+    Deoptimization::print_deoptimization_count();
+  }
+
   if (PrintMethodFlushingStatistics) {
     NMethodSweeper::print();
   }
