--- old/src/share/vm/interpreter/invocationCounter.hpp	2016-08-15 16:43:31.592545495 +0200
+++ new/src/share/vm/interpreter/invocationCounter.hpp	2016-08-15 16:43:31.500545495 +0200
@@ -41,6 +41,7 @@
 class InvocationCounter VALUE_OBJ_CLASS_SPEC {
   friend class VMStructs;
   friend class ciReplay;
+  friend class ciCacheReplay;
  private:                             // bit no: |31  3|  2  | 1 0 |
   unsigned int _counter;              // format: [count|carry|state]
 
