--- old/src/share/vm/opto/compile.cpp	2016-08-15 16:52:13.980545495 +0200
+++ new/src/share/vm/opto/compile.cpp	2016-08-15 16:52:13.888545495 +0200
@@ -26,6 +26,8 @@
 #include "asm/macroAssembler.hpp"
 #include "asm/macroAssembler.inline.hpp"
 #include "ci/ciReplay.hpp"
+#include "ci/ciCacheReplay.hpp"
+#include "ci/ciCacheProfiles.hpp"
 #include "classfile/systemDictionary.hpp"
 #include "code/exceptionHandlerTable.hpp"
 #include "code/nmethod.hpp"
@@ -876,6 +878,13 @@
   if (directive->DumpInlineOption && (ilt() != NULL)) {
     env()->dump_inline_data(_compile_id);
   }
+  // Dump profile to allow profile caching
+  if(env()->comp_level()>CompLevel_limited_profile && env()->comp_level() >= DumpProfilesMinTier) {
+    if ((DumpProfiles || method()->has_option("DumpProfile")) && (!method()->has_option("IgnoreDumpProfile"))) {
+		//tty->print("###Dump: %s\n",method()->holder()->name()->as_utf8());
+		_env->dump_cache_profiles(0, method()->name()->as_utf8());
+    }
+  }
 
   // Now that we know the size of all the monitors we can add a fixed slot
   // for the original deopt pc.
@@ -4207,6 +4216,8 @@
 void Compile::dump_inline_data(outputStream* out) {
   InlineTree* inl_tree = ilt();
   if (inl_tree != NULL) {
+    //tty->print(">DUMP_INLINE_DATA FOR:");method()->print_name(tty);tty->print("\n");
+    //tty->print(">INL_TREE->COUNT() = %d<\n",inl_tree->count());
     out->print(" inline %d", inl_tree->count());
     inl_tree->dump_replay_data(out);
   }
