--- old/src/share/vm/ci/ciClassList.hpp	2016-08-15 12:19:41.185019658 +0200
+++ new/src/share/vm/ci/ciClassList.hpp	2016-08-15 12:19:41.001019658 +0200
@@ -109,6 +109,8 @@
 friend class ciObjArray;               \
 friend class ciMetadata;               \
 friend class ciReplay;                 \
+friend class ciCacheProfiles;          \
+friend class ciCacheReplay;            \
 friend class ciTypeArray;              \
 friend class ciType;                   \
 friend class ciReturnAddress;          \
