--- old/src/share/vm/ci/ciEnv.hpp	2016-08-15 12:26:08.617019658 +0200
+++ new/src/share/vm/ci/ciEnv.hpp	2016-08-15 12:26:08.473019658 +0200
@@ -459,11 +459,16 @@
   void metadata_do(void f(Metadata*)) { _factory->metadata_do(f); }
 
   // Dump the compilation replay data for the ciEnv to the stream.
+  static bool first_dump;
   void dump_replay_data(int compile_id);
   void dump_inline_data(int compile_id);
   void dump_replay_data(outputStream* out);
   void dump_replay_data_unsafe(outputStream* out);
   void dump_compile_data(outputStream* out);
+  // Dump the cached profiling information for the ciEnv to the stream.
+  void dump_cache_profiles(int compile_id, const char* methodName);
+  void dump_cache_profiles(outputStream* out);
+  void dump_cache_profiles_unsafe(outputStream* out);
 };
 
 #endif // SHARE_VM_CI_CIENV_HPP
